src/input/Mouse.cc

```diff
@@ -239,10 +241,14 @@ void Mouse::write(uint8_t value, EmuTime::param time)
 				// sdsnatcher's post of 30 aug 2018 for a
 				// motivation for this difference:
 				//   https://github.com/openMSX/openMSX/issues/892
-				xRel = std::clamp(curXRel, -127, 127);
-				yRel = std::clamp(curYRel, -127, 127);
-				curXRel -= xRel;
-				curYRel -= yRel;
+				// Only introduce the next delta when we had a skipped xLow reading
+				// This to avoid trackball detection in the alternate cycle
+				if (xLow) {
+					xRel = std::clamp(curXRel, -127, 127);
+					yRel = std::clamp(curYRel, -127, 127);
+					curXRel -= xRel;
+					curYRel -= yRel;
+				} else xRel = yRel = 0;
```

Comment:
Sorry I don't understand this change. Maybe it is correct, but then it needs more comments.

This is my current understanding:
* In the normal case (when there's no timeout) the variable `phase` cycles through these 4 values `XHIGH`, `XLOW`, `YHIGH` and `YLOW` (in this order).
* When the `XHIGH` state is entered, we latch the variables `xRel` and `yRel`, these are the values that will be returned on reading (in the `read()` method).
* But when there is no timeout, the variable `xLow` can only be `true` in the phase `XHIGH` or `XLOW`. And in that case in this code block (when changing from `YLOW` to `XHIGH` we will always execute the `else` branch, and thus always set `xRel = yRel = 0`.
* This means that if a MSX program constantly reads the mouse (I mean more than once per frame, so not waiting for the timeout) it will effectively see no mouse movement. It's not impossible, but I strongly doubt that's the case with real MSX hardware.

Also the real MSX mouse hardware does not know when the MSX software actually reads from it (it constantly puts a high or low signal on the joystick port input pins (input-direction = from mouse to MSX), but it has no idea when the AY8910 chip actually 'uses' those signals). 

That means the real MSX mouse hardware cannot change behavior on whether or not the data was read. Maybe an alternative could be to do something (on line 215) when the timeout triggers. Real hardware can measure the time between *changes* in the output pins (output-direction = from MSX to mouse) (it cannot detect all writes, only writes that actually make a change).

So to summarize. probably this PR does make a good fix. But for sure it does it in a different way than how a real MSX mouse behaves. And I do believe this difference may break other MSX programs (e.g. programs that read in a tight loop).

@sdsnatcher73 Can you comment on this? Do you happen to know (in detail) how the MSX mouse protocol works, and how it interacts with trackball detection?

Reply:
Ok, here is what I have seen for mouse scanning in every case (I assume through BIOS):
write xHigh, read xHigh
write xLow, read xLow
write yHigh, read yHigh
write yLow, read yLow
write xHigh, read xHigh
write xLow
write yHigh
write yLow
[next cycle]

So my change will make sure that we only set the next delta (potentially large, which will never happen in HW, only with a virtual mouse) upon entering the xHigh state at the beginning of the cycle, and not in the middle.  In the middle the check happens for trackball.  So if the middle xHigh read is high, clipping to within trackball range happens in the next cycle.  Is that more clear now?

I never observed timeout BTW.  I guess it could happen, and I confess not having thought about what the effect will be, so I could think about that.  But overall, I do think this the right way to do this.  Remember that we also clamp between -127 and +127, not because the HW does it, but because our virtual mouse situation demands it.  This would be a similar fix ...

Reply:
Ok thanks, this part is becoming clear to me.
First of all: I think this is a separate (orthogonal) issue than the change in `InputEventGenerator.cc`, right?
* 1st part is about the (windows) host mouse grab handling.
* 2nd part is about msx mouse vs trackball detection confusion

It may be the case that the 2nd part now triggers more frequently(*), but it was in fact an already existing problem.

(*) Or actually does part 1 influence part 2? I wouldn't expect that. Warping the host mouse cursor should not influence what the MSX mouse sees. Or do I miss something?

In any case I believe it's better to split this over two commits (maybe you had done this initially, but were then asked to cleanup your commits and merge them, sorry if that's the case).

--

I did some research on the MSX mouse protocol, and how the MSX BIOS (or other MSX software) handles the mouse. It's hard to find good documentation. This [MRC wiki article](https://www.msx.org/wiki/Mouse/Trackball) is the best I could find.
* At the hardware level, the MSX mouse is close to what we emulate in openMSX. It has indeed 4 phases, with a reset to the 1st phase after some timeout.
* MSX software that directly accesses the hardware (I mean reading the mouse without using the BIOS) indeed closely follows this protocol. So just cycling through these 4 phases, thus reading 4 nibbles.
* What's new info (at least to me) is that the MSX BIOS reads **2 times** 4 phases. The expectation is that when you read the mouse a 2nd time (very close in time after the 1st read) then the real mouse will not have had the time to move very far(*2), and then you'll read an offset equal to zero. (An MSX trackball will instead reset the counter to -8 (no movement is represented by -8 instead 0), and that allows to distinguish between a mouse or a trackball).
* In openMSX this assumption breaks because we try to "smear" very large offsets over multiple successive cycles (with cycle I mean a group of 4 phases).

(*2) this "very far" part isn't true for negative offsets of course, also the smallest negative offset already has a non-zero upper nibble.

You rightfully claim that the real mouse does not have the smearing/clamping behavior, and that's indeed the root cause of breaking the detection heuristic. We did introduce this extra behavior to solve [this issue](https://github.com/openMSX/openMSX/issues/892). So I would like to keep this.

In this patch you try to detect the 2nd cycle, and then don't apply the smearing. I think that's a good solution, I only like to change the trigger for when to apply this. (Because as we discussed above, the current code may break MSX software that doesn't use the MSX BIOS to read the mouse).

The first idea that comes to my mind is:
* Instead of having 4 phases have 8 phases. So split `XHIGH` into `XHIGH1` and `XHIGH2` etc.
* These 4 extra phases behave identical to their original counterpart. So we don't change the MSX mouse protocol.
* Except for the *extra* smearing behavior that we added (that is not part of the actual mouse protocol), that will only be applied in `YLOW2`, not in `YLOW1`. (YLOW**2** is not a typo, it must be applied in the phase that comes before `XHIGH1`).
* The timeout will also reset to phase `YLOW2`, that way MSX software that doesn't use the BIOS (and assuming there's sufficient time between two read cycles), will always have the extra smearing heuristic.

Reply:
Ok, so yes, this approach with 8 phases works fine.  I have to clarify that I never saw timeouts.  What I should have said, that I never saw timeouts which had to change phase, i.e. the always happened when already in YLOW.  Setting it to YLOW2 will again normally not change the phase for BIOS driven mouse scans, but when this is bypassed, it can now typically timeout from YLOW1 to YLOW2, which should **theoretically** work fine (is there a known piece of MSX software that wants to read the mouse itself?).

I was able to link to this discussion in a comment in the code.  I have also made the variable initializations for the other point as you requested (I hope), but was not yet successful in solving the mouse warping issue (though it looks more and more as yet another SDL bug unfortunately).  I will see later what I can do.  For now I wanted to force push the code changes already, so that you can have a look at them.

You are right that these are in fact to different issues.  The relative mode issue and the mouse smearing out issue.  And indeed, the latter was always present but showed up clearly when I first try to use the alternate way of enabling relative mode (without automatic warping of the mouse).  It gave me virtual barriers.  However, now that I warp the mouse myself, the other issue is again less pressing, but of course we want to solve it, now that we know of it.  So, I will split both in two commits later.  For now let's live with one commit ...

Reply:
You could try SymbOS. I'm certain it does not use the BIOS to read out the mouse.l

Reply:
Yes, SymbOS works fine.  But since its mouse is slower (seen that with other MSX programs) a 10 pixel border is not enough.  Think about warping the mouse to the middle from all directions.  Need to test further the interaction of that method with the menubar ...